### PR TITLE
Define algorithm for determining whether high priority work exists.

### PR DESCRIPTION
**Pull request by rmcilroy**
Addresses #86 and #87.

### CODE REVIEW

**Comment by rmcilroy** [comment]
@rniwa, @yoavweiss, PTAL, thanks.

**Comment by rmcilroy** [comment]
Ping please

File: index.html [hunk 1]
```diff
@@ -465,8 +468,10 @@ <h2>The <code><dfn>IdleDeadline</dfn></code> interface</h2>
         <a>IdleDeadline</a> object.</li>
         <li>Let <var>timeRemaining</var> be <var>deadline</var> -
         <var>now</var>.</li>
-        <li>If the user agent now has a time-critical task pending,
-        set <var>timeRemaining</var> to 0.</li>
+        <li>Call the <a>is high-priority work pending algorithm</a>, passing
+	the <a>event loop</a> associated with <var>window</var> as
```
Comment (yoavweiss) [inline index.html:474]: The use of `window` here as a "global variable" seems a bit weird, but seems like this is how we're using it all over the place, so 🤷‍♂

**Review by yoavweiss — APPROVED** [review]
LGTM, but want @rniwa to make sure I'm not missing anything

**Comment by rmcilroy** [comment]
@rniwa could you have a look please. Does this address the issues you had?

File: index.html (hunk 1 above)
Comment (rniwa) [reply index.html:474]: We need to fix this. `window` here isn't well defined. Each instance of IdleDeadline should probably be associated with an event loop, or the window object. Then this part should refer to that associated event loop / window object's event loop.

File: index.html [hunk 2]
```diff
@@ -608,6 +614,18 @@ <h2>Invoke idle callback timeout algorithm</h2>
         </li>
       </ol>
     </section>
+    <section>
+      <h2>Is high-priority work pending algorithm</h2>
+      <p>The <dfn>is high-priority work pending algorithm</dfn>:</p>
```
Comment (rniwa) [inline index.html]: This should define arguments. e.g. <dfn>is high-priority work pending algorithm</dfn> given <var>event loop</var>

File: index.html [hunk 3]
```diff
@@ -465,8 +468,10 @@ <h2>The <code><dfn>IdleDeadline</dfn></code> interface</h2>
         <a>IdleDeadline</a> object.</li>
         <li>Let <var>timeRemaining</var> be <var>deadline</var> -
         <var>now</var>.</li>
-        <li>If the user agent now has a time-critical task pending,
-        set <var>timeRemaining</var> to 0.</li>
+        <li>Call the <a>is high-priority work pending algorithm</a>, passing
+	the <a>event loop</a> associated with <var>window</var> as
+	<var>event_loop</var>. If it returns <code>true</code> set
```
Comment (rniwa) [inline index.html]: We don't typically use underscored names for variable names in spec algorithms.

File: index.html [hunk 4]
```diff
@@ -608,6 +614,18 @@ <h2>Invoke idle callback timeout algorithm</h2>
         </li>
       </ol>
     </section>
+    <section>
+      <h2>Is high-priority work pending algorithm</h2>
+      <p>The <dfn>is high-priority work pending algorithm</dfn>:</p>
+      <ol>
+        <li>If <var>event_loop</var> has any task queues that contain a 
+	<a>task</a> <a>queued</a> with the <a>user interaction task source</a>,
```
Comment (rniwa) [inline index.html]: We need to restrict this to [*runnable* tasks](https://html.spec.whatwg.org/multipage/webappapis.html#concept-task-runnable).

Each [task](https://html.spec.whatwg.org/multipage/webappapis.html#concept-task) is a struct which has a source as its member so there is no need to say a task queue contains a task queued with a particular task source. use say a task queue contains a task with a particular source; e.g. if <var>event loop</var> has any task queues with a runnable task with <a>user interaction task source</a>.

But do we really want to restrict this to just "user interaction task source"? Networking tasks are probably important and so are tasks with [media element event ask sources](https://html.spec.whatwg.org/multipage/media.html#media-element-event-task-source).

File: index.html [hunk 5]
```diff
@@ -608,6 +614,18 @@ <h2>Invoke idle callback timeout algorithm</h2>
...
+	or if the user agent determines that it MAY soon enqueue such a task,
+        then return <code>true</code>.
+	<li>If the user determines that it has internal high-priority work
```
Comment (rniwa) [inline index.html]: You mean the user *agent*?

**Comment by rniwa** [comment]
Note the PR as it stands excludes any task enqueued that's not of user interaction source because the current wording is that "If the user [agent ]determines that it has internal high-priority work pending, then return true", thereby excluding any tasks that have already been enqueued.

It might be better to simply say that "[i]f the user agent dependent mechanism determines that there is a higher priority work than executing idle callbacks, return true".

**Comment by rmcilroy** [comment]
> Note the PR as it stands excludes any task enqueued that's not of user interaction source

Yes that is intentional. The intention is to only reset the deadline if there is user input task pending (with a similar algorithm to the is-input-pending proposal). The idle task won't start if there are any other task pending (since it's queued on the lowest priority idle task queue), but once it's started we only want it to yield earlier than its original deadline if really high priority (e.g., user input) events occur.

**Comment by rniwa** [comment]
How about rAF? Shouldn't we yielding once the rendering opportunity arrives?

**Comment by rmcilroy** [comment]
That should already be taken into account when computing the deadline initially.

**Comment by rniwa** [comment]
> That should already be taken into account when computing the deadline initially.

As I have stated previously, this is not possible in WebKit. We can't precisely determine when a rendering opportunity would occur. Given this, it seems prudent that this definition will take that into account.

**Comment by rmcilroy** [comment]
I thought we agreed during the call that WebKit could do a similar estimation as we do in Chrome? In our case the renderer process also doesn't control rAF trigger, the browser process signals it to us. We estimate next rAF based on current frame intervals. In the notes it's recorded as this:

> Ryosuke: ... we can’t query raF, that’s owned by core animation
> Ross: we kinda have similar problem, we rely on browser process to tell us
> … we know what is expected, we speculate based on frame frequency
> Ryosuke: we can also do that, I guess

Either way, it is important that the initial deadline takes rAF into account as this is one of the main use-cases of rIC (enabling the JS devs to estimate how much time they have to do work before the next frame is scheduled.

**Comment by rniwa** [comment]
> I thought we agreed during the call that WebKit could do a similar estimation as we do in Chrome?

I never agreed to do anything. Only thing I said is that we could do something like that, not that we're going to do it or that it's something we want to implement. Furthermore, because of the way things are architected, such an estimate would be very much inaccurate in WebKit even if we were to implement such an estimate.

> Either way, it is important that the initial deadline takes rAF into account as this is one of the main use-cases of rIC (enabling the JS devs to estimate how much time they have to do work before the next frame is scheduled.

Well, an estimation is one thing. Knowing it has come is another. There is no reason rIC can't yield once the rendering opportunity has occurred to a document.

**Comment by rmcilroy** [comment]
> I never agreed to do anything. Only thing I said is that we could do something like that, not that we're going to do it or that it's something we want to implement. 

I was not trying to imply that you'd agreed to implementing this in Webkit, merely that we had discussed it as being possible - apologies if it came across that way.

> Well, an estimation is one thing. Knowing it has come is another. There is no reason rIC can't yield once the rendering opportunity has occurred to a document.

We can certainly add an item checking if there is a rendering opportunity here (I'll add it to the PR), but this should only really trigger in infrequent cases where something happened that we couldn't predict at deadline calculation time (e.g., frame rate changed or JS triggered a rendering opportunity we didn't know about before). 

The initial deadline should have baked in an estimate of when the next rendering opportunity is likely to start so that it can be used by the rIC callback user-code to make an ahead-of-time decision of whether to start long-running work or wait for a future idle period that might have a longer deadline available.

File: index.html (hunk 4 above)
Comment (rmcilroy) [reply index.html]: Done

File: index.html (hunk 3 above)
Comment (rmcilroy) [reply index.html]: Done

File: index.html (hunk 2 above)
Comment (rmcilroy) [reply index.html]: Done

File: index.html (hunk 1 above)
Comment (rmcilroy) [reply index.html:474]: Done

File: index.html (hunk 5 above)
Comment (rmcilroy) [reply index.html]: Yes thanks. Done.

**Comment by rmcilroy** [comment]
Comments should be addressed, please take another look @rniwa

**Comment by rmcilroy** [comment]
Are we good to land this?

**Comment by yoavweiss** [comment]
@rniwa - friendly ping! :)

**Comment by rmcilroy** [comment]
@annevk could you please review this PR? Unfortunately Ryosuke doesn't have bandwidth to do this review right now.

File: index.html [hunk 6]
```diff
@@ -550,17 +559,20 @@ <h2>Start an idle period algorithm</h2>
       <h2>Invoke idle callbacks algorithm</h2>
       <p>The <dfn>invoke idle callbacks algorithm</dfn>:</p>
       <ol>
-        <li>If the user-agent believes it should end the idle period early due
-        to newly scheduled high-priority work, skip to step 4.
+        <li>Call the <a>is high-priority work pending algorithm</a>, passing
+	the <a>event loop</a> associated with <var>window</var> as
+	<var>eventLoop</var>. If it returns <code>true</code>, skip to step 4.
```
Comment (annevk) [inline index.html:564]: Can we avoid goto?

File: index.html [hunk 7]
```diff
@@ -599,15 +611,32 @@ <h2>Invoke idle callback timeout algorithm</h2>
...
       </ol>
     </section>
+    <section>
+      <h2>Is high-priority work pending algorithm</h2>
+      <p>The <dfn>is high-priority work pending algorithm</dfn> given
+      <var>eventLoop</var>:</p>
+      <ol>
+        <li>If <var>eventLoop</var> has any <a>task queues</a> with a
+        <a>runnable task</a> with the <a>user interaction task source</a>,
+        or if the user agent determines that it MAY soon enqueue such a task,
```
Comment (annevk) [inline index.html:632]: I suspect you want "might" here as this is not a conformance requirement.

**Review by annevk — COMMENTED** [review]
This is really hard to review without having the generated HTML or a generated HTML diff. Are those available? I'm also not really familiar with the details here.

Perhaps @farre has some ideas.

Left a couple minor nits nevertheless. Hope they help.

**Comment by yoavweiss** [comment]
Obsoleted by #95

**Pull request closed without merging**